src/Microsoft.NetCore.Analyzers/Core/Security/UseXmlReaderForSchemaRead.cs

```diff
@@ -0,0 +1,112 @@
+                DiagnosticId,
+                s_Title,
+                s_Message,
+                DiagnosticCategory.Security,
+                DiagnosticHelpers.DefaultDiagnosticSeverity,
+                isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+                description: s_Description,
+                helpLinkUri: null,
+                customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+
+            // Security analyzer - analyze and report diagnostics on generated code.
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterCompilationStartAction(compilationStartAnalysisContext =>
+            {
+                var compilation = compilationStartAnalysisContext.Compilation;
+                var dataSetTypeSymbol = compilation.GetTypeByMetadataName(WellKnownTypeNames.SystemDataDataSet);
+
+                if (dataSetTypeSymbol == null)
+                {
+                    return;
+                }
+
+                var xmlReaderTypeSymbol = compilation.GetTypeByMetadataName(WellKnownTypeNames.SystemXmlXmlReader);
+
+                compilationStartAnalysisContext.RegisterOperationAction(operationAnalysisContext =>
+                {
+                    var invocationOperation = (IInvocationOperation)operationAnalysisContext.Operation;
+                    var methodSymbol = invocationOperation.TargetMethod;
+                    var methodName = methodSymbol.Name;
+
+                    if (methodName.StartsWith("ReadXml", StringComparison.Ordinal) &&
+                        IsOverrides(methodSymbol, dataSetTypeSymbol))
+                    {
+                        if (xmlReaderTypeSymbol != null &&
+                            methodSymbol.Parameters.Length > 0 &&
+                            methodSymbol.Parameters[0].Type.Equals(xmlReaderTypeSymbol))
+                        {
+                            return;
+                        }
+
+                        operationAnalysisContext.ReportDiagnostic(
+                            invocationOperation.CreateDiagnostic(
+                                Rule,
+                                methodName));
+                    }
+                }, OperationKind.Invocation);
+
+                /// <summary>
+                /// Find out if the method overrides from target virtual method of a certain class
+                /// </summary>
+                /// <param name="methodSymbol">The method</param>
+                /// <param name="classSymbol">The class has virtual method</param>
+                bool IsOverrides(IMethodSymbol methodSymbol, INamedTypeSymbol classSymbol)
```

Comment:
Are you probably going to use this same code later?  Might be worth putting it inside of IMethodSymbolExtensions.cs.